File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.cc
#include <iostream>
#include <queue>
#include <tuple>
using namespace std;

char board[1000][1001];
int dist[1000][1000][2];
int dx[] = {1, 0, -1, 0};
int dy[] = {0, 1, 0, -1};

int main() {
    ios::sync_with_stdio(0); cin.tie(0);
    int N, M; cin >> N >> M;
    for(int i=0; i<N; i++) cin >> board[i];

    int ans = -1, endX = N-1, endY = M-1;
    queue<tuple<int, int, bool>> Q;
    dist[0][0][0] = dist[0][0][1] = 1;
    Q.push({0, 0, 0});
    while(!Q.empty()) {
        int x, y; bool broken;
        tie(x, y, broken) = Q.front(); 
        if(x == endX && y == endY) {
            ans = dist[x][y][broken];
            break;
        }
        Q.pop();
        int nxt = dist[x][y][broken] + 1;
        for(int d=4; d--;) {
            int nx = x + dx[d], ny = y + dy[d];
            if(nx<0 || nx >=N || ny<0 || ny>=M) continue;
            if(board[nx][ny] == '0') {
                if(dist[nx][ny][broken]) continue;
                dist[nx][ny][broken] = nxt;
                Q.push({nx, ny, broken});
            }
            if(board[nx][ny] == '1' && !broken && !dist[nx][ny][1]) {
                dist[nx][ny][1] = nxt;
                Q.push({nx, ny, 1});
            }
        }
    }
    cout << ans;
}

File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.py
import sys
from collections import deque
readline = sys.stdin.readline

dxy = ((1,0), (-1,0), (0,1), (0,-1))

N, M = map(int, readline().split())
board = [readline().strip() for _ in range(N)]
dist = [[[0, 0] for _ in range(M)] for _ in range(N)] 
# dist[N][M][2]

ans = -1
dist[0][0][0] = dist[0][0][1] = 1
queue = deque([(0, 0, 0)])

while queue:
    x, y, broken = queue.popleft()
    if (x, y) == (N - 1, M - 1):
        ans = dist[x][y][broken]
        break
      
    nd = dist[x][y][broken] + 1
    for dx, dy in dxy:
        nx, ny = x + dx, y + dy
        if not (0 <= nx < N and 0 <= ny < M):
            continue
          
        if board[nx][ny] == '0' and dist[nx][ny][broken] == 0:
            dist[nx][ny][broken] = nd
            queue.append((nx, ny, broken))
        elif board[nx][ny] == '1' and broken == 0 and dist[nx][ny][1] == 0:
            dist[nx][ny][1] = nd
            queue.append((nx, ny, 1))

print(ans)